libvast/src/system/source.cpp

```diff
@@ -168,10 +190,10 @@ source(caf::stateful_actor<source_state>* self, format::reader_ptr reader,
   self->state.table_slice_size = table_slice_size;
   self->state.has_sink = false;
   self->state.done = false;
-  self->state.transformer
-    = self->spawn(transformer, "source-transformer", std::move(pipelines));
-  if (!self->state.transformer) {
-    VAST_ERROR("{} failed to spawn transformer", *self);
+  self->state.executor = pipeline_executor{std::move(pipelines)};
+  if (auto err = self->state.executor.validate(
+        pipeline_executor::allow_aggregate_pipelines::yes)) {
+    VAST_ERROR("source contains invalid pipeline {}", err);
     self->quit();
```

Comment:
One minor design note on this while I'm seeing it (even though you didn't this logic): In general we always want to print errors as late as possible because that allows us to either handle the error, or to further enrich it with context before printing it.

The pattern here will lead to a VAST client that exits with a zero exit code, indicating a successful exit, because the error isn't "returned" from the actor. To achieve that, you usually want to do the following instead:

```c++
self->quit(caf::make_error(...));
return my_actor::behavior_type::make_empty_behavior(); // or just {} for untyped actors
```

Actors that monitor the source (i.e., the import command's scoped actor) will then receive a down message from the source with exactly this error, and can render it before shutting down.

Reply:
Perfect, thanks!